src/content/posts/2021/pitfalls-of-unoptimzied-nuget-packages/index.md

```diff
@@ -0,0 +1,177 @@
+</div>
+<br>
+{{< /rawhtml >}}
+
+## Anything that can go wrong will go wrong - [Kristian Murphy Hellang](https://twitter.com/khellang)
+
+I feel like publishing packages to _nuget.org_ relates a lot to sailing and running aground:
+
+> There are three kinds of skippers, those who have run aground, those who will run aground, and those that have but won't admit it.
+
+Replace _skippers_ with _package authors_ and _running aground_ with _publish packages built in debug_ and there you have it.
+
+Publishing unoptimized bits also happened to Sentry's _SharpRaven_ SDK back in 2018. If you're using [SharpRaven version 2.3.0](https://www.nuget.org/packages/SharpRaven/2.3.0), you're running an unoptimized assembly. I would recommend you [upgrade it to 2.3.1](https://github.com/getsentry/raven-csharp/pull/226/files) but on that same year that package was replaced by [Sentry](https://www.nuget.org/packages/Sentry/). Don't worry, it's [packaged in release mode](https://github.com/getsentry/sentry-dotnet/blob/bc4c5e23eaaaa848a9a44ae05ac257a591624b24/build.ps1#L14). Not that you should take my word for it.
+
+In early 2020, I even [opened a PR to add this verification to the Sentry SDK for .NET](https://github.com/getsentry/sentry-dotnet/pull/365/files) so it would check assemblies being loaded whether they were optimized or not, and send events to Sentry when applicable. But it never felt like the right thing to do. This verification must happen much earlier, at latest during build time.
+
+## Release build and compiler optimization
+
+I've been talking about packaging NuGet with `Release` configuration then talking about Roslyn optimization.
+These two things are not exactly the same (though by default related) and I wish to clarify.
+
+The `Release` configuration, by default does provide the `-optimize+` flag to the C# compiler 
+but you can achieve the same for example by means of a MSBuild property `<Optimize>true</Optimize>`.
+I hope the references to the docs above already made this clear.
+
+## My days are numbered
+
+That's a gloomy heading. It's the `UnoptimizedAssemblyDetector` NuGet package talking though, not me.
+I built this package in hopes for it to become obsolete one day. Not because I expect the default of `dotnet pack` to ever change.
+Clearly it won't. But I believe such warnings should be part of NuGet itself.
+
+When you push a NuGet package to _nuget.org_, it should verify the managed assemblies under `lib/` are indeed compiled with optimization, and if not, it should warn the author.
+At the same time, when restoring such package, it should warn the consumer. This part in particular will make `UnoptimizedAssemblyDetector` redundant.
+
+What else can **we** do? [NuGet Trends](https://github.com/dotnet/nuget-trends) has the complete catalog. It could be extended to download the packages and inspect the assemblies in them. With that, create graphs and expose an endpoint to list unoptimized assemblies.
```

Comment:
It sounds like you're going out of your way to make sure the reader understands that the impact of unoptimized assemblies is great (otherwise why would you be concerned about it so much so as to list multiple approaches to avoid this problem?). But as a reader, it's not clear why because the evidence was never presented.

Reply:
Showing side by side IL with `nop` vs without or some inlining would convince u? Or benchmarks?

Then one could say benchmarks are not representational. And it all depends on the app itself. And in some cases doesn't really matter as it's not noticeable. So it's a losing batter.

I would like to assume the reader already accepts the fact it's relevant. If not they are not the audience I'm looking for (they'll be the ones writing negative things like "Doesn't make a difference" on reddit later) :)